### PR TITLE
feat: add RAG model settings

### PR DESCRIPTION
**Pull request by ram02z**
## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Tests

https://trello.com/c/mU1rkghL/117-add-similarity-cutoff-using-rag-retrieval
https://trello.com/c/NXzBoM4M/118-allow-top-k-setting-for-rag-retrieval

## Proposed Changes

This PR aims to introduce two RAG settings: distance threshold and top-k. The defaults are 0.5 and 1, respectively. The web app includes the settings in a collapsible section.

## Screenshots

image

### CODE REVIEW

**Comment by jiaweihong** [comment]
Just a heads up, the create_collection func from chromadb requires a collection name to be between 3-63 characters. I don't think it will affect us currently since we are giving it a default name, but if we plan to start creating different collections for different uploads, we should be aware of it.

Line 133
Screenshot 2024-04-04 at 2 14 26 PM

**Comment by jiaweihong** [comment]
Also just so I'm understanding this correctly, the way the collections lifecycle works is that everytime the api is called, we want a fresh collection to be created right? So for every api call, we essentially have a brand new collection

**Comment by jiaweihong** [comment]
Screenshot 2024-04-04 at 2 26 44 PM
We are setting the max token limit in the function definition to be 512


Screenshot 2024-04-04 at 2 27 33 PM
but when we call the function we pass a default value of 32, why not just set it to be in 32 in the function dev

File: src/alinet/rag/db.py [hunk 1]
```diff
@@ -127,22 +136,30 @@ def create_collection(self, collection_name: str = "default"):
...
 
-        relevant_query = collection.query(query_embeddings=query_embedding, n_results=1)
-
-        relevant_context = relevant_query["documents"][0][0]
+        relevant_queries = collection.query(
+            query_embeddings=query_embedding, n_results=top_k
+        )
 
-        long_answer_with_relevant_context = f"{context} {relevant_context}"
+        for i in range(top_k):
```
Comment (wilsvd) [inline src/alinet/rag/db.py:152]: We get an indexing error. Here are some screenshots.
image

image

```print("Relevant Query:")
        pprint(relevant_queries)
        print("Number of relevant queries")
        pprint(len(relevant_queries["distances"]))
        for i in range(top_k):
            if relevant_queries["distances"][i][0] > distance_threshold:
                relevant_context = relevant_queries["documents"][i][0]

                context = f"{context} {relevant_context}"
        print()
```

```python
  File "/home/wilvdh/mmqg/src/alinet/rag/db.py", line 158, in add_relevant_context_to_source
    if relevant_queries["distances"][i][0] > distance_threshold:
IndexError: list index out of range
```

Comment (wilsvd) [reply src/alinet/rag/db.py:152]: The distances is 2D array and here its length is 1. The inside array has a length of top_k. But we are trying to access the index in the outer array.

E.g. we have, [[x, y, z]] is length of 1
We trying to do arr[1][0] but arr[1] doesn't exist, we only have arr[0][0..top_k]

**Review by wilsvd — CHANGES_REQUESTED** [review]
Other than the indexing issue on the backend, the frontend logic seems fine.

File: src/alinet/rag/db.py (hunk 1 above)
Comment (ram02z) [reply src/alinet/rag/db.py:152]: Ah yeah, my bad. the upper limit should be the length of the array not top-k.

**Comment by jiaweihong** [comment]
Screenshot 2024-04-04 at 3 02 29 PM
If top_k is more than 1, in the current implementation, it will just add the last index of top_k but we want it to add, im assuming all top k documents so it should be +=

**Comment by ram02z** [comment]
> If top_k is more than 1, in the current implementation, it will just add the last index of top_k but we want it to add, im assuming all top k documents so it should be += 

`context = f"{context} {relevant_context}"` is the same as `context += f" {relevant_context}"`

**Comment by jiaweihong** [comment]
oh right yeah im retarded

**Review by wilsvd — APPROVED** [review]
lgtm

**Pull request merged**